Microsoft.Toolkit/Diagnostics/Generated/TypeInfo.ttinclude

```diff
@@ -0,0 +1,96 @@
+﻿<#@ template language="C#" #>
+<#@ assembly name="System.Core" #>
+<#@ import namespace="System.Collections.Generic" #>
+<#+
+    /// <summary>
+    /// A model representing the info on an enumerable type
+    /// </summary>
+    sealed class EnumerableTypeInfo
+    {
+        public EnumerableTypeInfo(
+            string type,
+            string xmlType,
+            string name,
+            string size,
+            string destinationType,
+            string cast)
+        {
+            Type = type;
+            XmlType = xmlType;
+            Name = name;
+            Size = size;
+            DestinationType = destinationType;
+            Cast = cast;
+        }
+
+        /// <summary>
+        /// Gets the name of the current type
+        /// </summary>
+        public string Type { get; }
+
+        /// <summary>
+        /// Gets the XML-formatted name of the current type (eg. with {T} instead of &lt;T>)
+        /// </summary>
+        public string XmlType { get; }
+
+        /// <summary>
+        /// Gets the variable name to use
+        /// </summary>
+        public string Name { get; }
+
+        /// <summary>
+        /// Gets the name of the property to use to retrieve the length of the current type
+        /// </summary>
+        public string Size { get; }
+
+        /// <summary>
+        /// Gets whether or not the current type has a "Count" property
+        /// </summary>
+        public bool HasCountProperty => Size == "Count";
+
+        /// <summary>
+        /// Gets the name of the destination type, when comparing counts across different collections
+        /// </summary>
+        public string DestinationType { get; }
+
+        /// <summary>
+        /// Gets the (optional) casting to resolve the diamond problem between different interfaces
+        /// </summary>
+        public string Cast { get; }
+    }
+
+    /// <summary>
+    /// Gets the list of available enumerable types to generate APIs for
+    /// </summary>
+    static readonly IReadOnlyList<EnumerableTypeInfo> EnumerableTypes = new[]
+    {
```

Comment:
Isn't `IEnumerable` missing here?

Reply:
I had it too initially but then I removed it because I didn't think it made sense. All the `Guard` APIs mostly use collections to check their size, and that's not possible with `IEnumerable<T>`. You'd need to create the enumerator and traverse the sequence, but that could get really expensive if the enumerator is doing some non-trivial work, so it's generally bad practice. By not having them we make the intent clearer and force users to first get an actual collection to work on, so that they skip multiple enumerations. I feel like that's one of the easiest mistakes to make especially for beginners, so I thought it would be a good idea to give them a hint by not supporting the type at all here.

Reply:
Cool, just wanted to double-check it wasn't overlooked :)